### PR TITLE
11-get /api/articles (topic query)

### PR DESCRIPTION
**Pull request by booayl**
Added following feature for endpoint: **GET /api/articles**:

- accept query: topic, which filters the articles by the topic value specified in the query. If the query is not mentioned, the endpoint should respond with all articles.

Happy Path: 
_**/api/articles?topic="topic_name"**_
GET200: Endpoint accept topic query and responds articles with selected topics

Sad Path: 
GET404: Respond with an error when passed non-existent topic
GET404: Respond with an error when passed in invalid query

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -351,9 +351,49 @@ describe("GET /api/users",()=>{
...
+  test("GET200: Endpoint accept topic query and responds articles with seleted topics", () => {
+    return request(app)
+      .get("/api/articles?topic=cats")
+      .expect(200)
+      .then(({ body: { allArticles } }) => {
+        expect(allArticles.length).toBe(1);
+        allArticles.forEach((article) => {
+          expect(article).toMatchObject({
+            title: expect.any(String),
+            topic: "cats",
```
Comment (roisinoreilly) [inline __tests__/app.test.js:371]: nicely done

File: __tests__/app.test.js [hunk 2]
```diff
@@ -351,9 +351,49 @@ describe("GET /api/users",()=>{
...
+  test("GET404: Respond with an error when passed non-existent topic", () => {
+    return request(app)
+      .get("/api/articles?topic=abc")
+      .expect(404)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Non-existent Topic");
+      });
+  });
+  test("GET404: Respond with an error when invalid query", () => {
```
Comment (roisinoreilly) [inline __tests__/app.test.js:390]: this should maaaybe be a 400 instead since it's an invalid request

File: model.js [hunk 3]
```diff
@@ -17,80 +17,114 @@ exports.fetchArticleById = (article_id) => {
...
+    if (rows.length === 0) {
+      return Promise.reject({ status: 404, msg: "Non-existent Topic" });
+    }
     return rows;
   });
 };
 
-exports.fetchCommentsByArticleID = (article_id,sort_by = "created_at", order = "DESC") =>{
-    let sqlQueryString = `SELECT comments.* 
+exports.validateQuery = (queries) => {
```
Comment (roisinoreilly) [inline model.js:44]: interesting way of doing it! I like that it's "future-proof" and you could potentially add new columns to the db and you won't need to change your query validation at all, really nicely done

File: controllers.js [hunk 4]
```diff
@@ -29,10 +30,16 @@ exports.getArticleById = (req, res, next) => {
...
 
 exports.getArticles = (req, res, next) => {
-  const { sort_by, order } = req.query;
-  fetchArticles(sort_by, order).then((articles) => {
-    res.status(200).send({ allArticles: articles });
-  });
+  const { sort_by, order, topic } = req.query;
+  const queries = Object.keys(req.query);
+
+  Promise.all([fetchArticles(sort_by, order, topic), validateQuery(queries)])
```
Comment (roisinoreilly) [inline controllers.js:36]: excellent

**Review by roisinoreilly — COMMENTED** [review]
excellent work on this one and congratulations for getting this far!! I especially like your query validation I think it's really clever and while it's complicated I think it is better than just hardcoding a list of valid queries. Have a think about the 404 error I commented on and what you think, but really well done!!

File: __tests__/app.test.js (hunk 2 above)
Comment (booayl) [reply __tests__/app.test.js:390]: Oops, it is 400 in the test, and should be GET400 in the description 😅 just updated it. 

I do struggle to decide to go for 400 or 404 every time, is there like a way to know which one I should use? Or is it different for every situation?

**Pull request merged**